Flaky test: TestCloseConnectionWhenServerPrefaceNotReceived

```
E0108 22:41:00.760194    6611 pickfirst.go:59] pickfirstBalancer: failed to NewSubConn: grpc: cannot create SubConn with empty address list
==================
WARNING: DATA RACE
Write at 0x000000e1a608 by goroutine 281:
  google.golang.org/grpc.TestCloseConnectionWhenServerPrefaceNotReceived()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn_test.go:178 +0xca
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c
Previous read at 0x000000e1a608 by goroutine 229:
  google.golang.org/grpc.(*addrConn).resetTransport()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:1068 +0x3ec
  google.golang.org/grpc.(*addrConn).transportMonitor()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:1273 +0x619
  google.golang.org/grpc.(*addrConn).connect.func1()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:837 +0x268
Goroutine 281 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:921 +0x206
  main.main()
      google.golang.org/grpc/_test/_testmain.go:220 +0x1d3
Goroutine 229 (finished) created at:
  google.golang.org/grpc.(*addrConn).connect()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:828 +0x185
  google.golang.org/grpc.(*acBalancerWrapper).Connect()
      /home/travis/gopath/src/google.golang.org/grpc/balancer_conn_wrappers.go:293 +0x88
  google.golang.org/grpc.(*pickfirstBalancer).HandleResolvedAddrs()
      /home/travis/gopath/src/google.golang.org/grpc/pickfirst.go:63 +0x4ab
  google.golang.org/grpc.(*ccBalancerWrapper).watcher()
      /home/travis/gopath/src/google.golang.org/grpc/balancer_conn_wrappers.go:139 +0x4e6
==================
--- FAIL: TestCloseConnectionWhenServerPrefaceNotReceived (3.50s)
	testing.go:699: race detected during execution of test
```

#1732 and #1748 both attempted to fix this already; I'm not sure if this is the same failure mode or not.

--- a/clientconn.go
+++ b/clientconn.go
@@ -926,7 +926,7 @@ func (cc *ClientConn) resolveNow(o resolver.ResolveNowOption) {
 
 // Close tears down the ClientConn and all underlying connections.
 func (cc *ClientConn) Close() error {
-	cc.cancel()
+	defer cc.cancel()
 
 	cc.mu.Lock()
 	if cc.conns == nil {

--- a/clientconn_test.go
+++ b/clientconn_test.go
@@ -187,7 +187,7 @@ func TestCloseConnectionWhenServerPrefaceNotReceived(t *testing.T) {
 	defer func() {
 		lis.Close()
 		// conn2 shouldn't be closed until the client has
-		// observerd a successful test.
+		// observed a successful test.
 		if conn2 != nil {
 			conn2.Close()
 		}

--- a/clientconn_test.go
+++ b/clientconn_test.go
@@ -176,19 +176,18 @@ func TestCloseConnectionWhenServerPrefaceNotReceived(t *testing.T) {
 	}()
 	defer leakcheck.Check(t)
 	minConnectTimeout = time.Millisecond * 500
-	server, err := net.Listen("tcp", "localhost:0")
+	lis, err := net.Listen("tcp", "localhost:0")
 	if err != nil {
 		t.Fatalf("Error while listening. Err: %v", err)
 	}
 	var (
-		conn1, conn2 net.Conn
-		over         uint32
+		conn2 net.Conn
+		over  uint32
 	)
 	defer func() {
-		server.Close()
-		if conn1 != nil {
-			conn1.Close()
-		}
+		lis.Close()
+		// conn2 shouldn't be closed until the client has
+		// observerd a successful test.
 		if conn2 != nil {
 			conn2.Close()
 		}
@@ -196,14 +195,14 @@ func TestCloseConnectionWhenServerPrefaceNotReceived(t *testing.T) {
 	done := make(chan struct{})
 	go func() { // Launch the server.
 		defer close(done)
-		var err error
-		conn1, err = server.Accept()
+		conn1, err := lis.Accept()
 		if err != nil {
 			t.Errorf("Error while accepting. Err: %v", err)
 			return
 		}
+		defer conn1.Close()
 		// Don't send server settings and the client should close the connection and try again.
-		conn2, err = server.Accept() // Accept a reconnection request from client.
+		conn2, err = lis.Accept() // Accept a reconnection request from client.
 		if err != nil {
 			t.Errorf("Error while accepting. Err: %v", err)
 			return
@@ -228,13 +227,13 @@ func TestCloseConnectionWhenServerPrefaceNotReceived(t *testing.T) {
 			break
 		}
 	}()
-	client, err := Dial(server.Addr().String(), WithInsecure())
+	client, err := Dial(lis.Addr().String(), WithInsecure())
 	if err != nil {
 		t.Fatalf("Error while dialing. Err: %v", err)
 	}
 	time.Sleep(time.Second * 2) // Let things play out.
 	atomic.StoreUint32(&over, 1)
-	server.Close()
+	lis.Close()
 	client.Close()
 	<-done
 }

--- a/clientconn_test.go
+++ b/clientconn_test.go
@@ -19,7 +19,6 @@
 package grpc
 
 import (
-	"io"
 	"math"
 	"net"
 	"sync/atomic"
@@ -173,99 +172,71 @@ func TestCloseConnectionWhenServerPrefaceNotReceived(t *testing.T) {
 	mctBkp := minConnectTimeout
 	// Call this only after transportMonitor goroutine has ended.
 	defer func() {
-		atomic.StoreInt64((*int64)(&minConnectTimeout), int64(mctBkp))
+		minConnectTimeout = mctBkp
 	}()
 	defer leakcheck.Check(t)
-	atomic.StoreInt64((*int64)(&minConnectTimeout), int64(time.Millisecond*500))
+	minConnectTimeout = time.Millisecond * 500
 	server, err := net.Listen("tcp", "localhost:0")
 	if err != nil {
 		t.Fatalf("Error while listening. Err: %v", err)
 	}
-	defer server.Close()
+	var (
+		conn1, conn2 net.Conn
+		over         uint32
+	)
+	defer func() {
+		server.Close()
+		if conn1 != nil {
+			conn1.Close()
+		}
+		if conn2 != nil {
+			conn2.Close()
+		}
+	}()
 	done := make(chan struct{})
-	clientDone := make(chan struct{})
 	go func() { // Launch the server.
-		defer func() {
-			if done != nil {
-				close(done)
-			}
-		}()
-		conn1, err := server.Accept()
+		defer close(done)
+		var err error
+		conn1, err = server.Accept()
 		if err != nil {
 			t.Errorf("Error while accepting. Err: %v", err)
 			return
 		}
-		defer conn1.Close()
-		// Don't send server settings and make sure the connection is closed.
-		time.Sleep(time.Millisecond * 1500) // Since the first backoff is for a second.
-		conn1.SetDeadline(time.Now().Add(time.Second))
-		b := make([]byte, 24)
-		for {
-			// Make sure the connection was closed by client.
-			_, err = conn1.Read(b)
-			if err == nil {
-				continue
-			}
-			if err != io.EOF {
-				t.Errorf(" conn1.Read(_) = _, %v, want _, io.EOF", err)
-				return
-			}
-			break
-		}
-
-		conn2, err := server.Accept() // Accept a reconnection request from client.
+		// Don't send server settings and the client should close the connection and try again.
+		conn2, err = server.Accept() // Accept a reconnection request from client.
 		if err != nil {
 			t.Errorf("Error while accepting. Err: %v", err)
 			return
 		}
-		defer conn2.Close()
 		framer := http2.NewFramer(conn2, conn2)
-		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+		if err = framer.WriteSettings(http2.Setting{}); err != nil {
 			t.Errorf("Error while writing settings. Err: %v", err)
 			return
 		}
-		time.Sleep(time.Millisecond * 1500) // Since the first backoff is for a second.
-		conn2.SetDeadline(time.Now().Add(time.Millisecond * 500))
+		b := make([]byte, 8)
 		for {
-			// Make sure the connection stays open and is closed
-			// only by connection timeout.
 			_, err = conn2.Read(b)
 			if err == nil {
 				continue
 			}
-			if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+			if atomic.LoadUint32(&over) == 1 {
+				// The connection stayed alive for the timer.
+				// Success.
 				return
 			}
 			t.Errorf("Unexpected error while reading. Err: %v, want timeout error", err)
 			break
 		}
-		close(done)
-		done = nil
-		<-clientDone
-
 	}()
 	client, err := Dial(server.Addr().String(), WithInsecure())
 	if err != nil {
 		t.Fatalf("Error while dialing. Err: %v", err)
 	}
-	<-done
-	// TODO: The code from BEGIN to END should be delete once issue
-	// https://github.com/grpc/grpc-go/issues/1750 is fixed.
-	// BEGIN
-	// Set underlying addrConns state to Shutdown so that no reconnect
-	// attempts take place and thereby resetting minConnectTimeout is
-	// race free.
-	client.mu.Lock()
-	addrConns := client.conns
-	client.mu.Unlock()
-	for ac := range addrConns {
-		ac.mu.Lock()
-		ac.state = connectivity.Shutdown
-		ac.mu.Unlock()
-	}
-	// END
+	time.Sleep(time.Second * 2) // Let things play out.
+	atomic.StoreUint32(&over, 1)
+	server.Close()
 	client.Close()
-	close(clientDone)
+	<-done
 }
 
 func TestBackoffWhenNoServerPrefaceReceived(t *testing.T) {

--- a/clientconn_test.go
+++ b/clientconn_test.go
@@ -22,6 +22,7 @@ import (
 	"io"
 	"math"
 	"net"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -172,10 +173,10 @@ func TestCloseConnectionWhenServerPrefaceNotReceived(t *testing.T) {
 	mctBkp := minConnectTimeout
 	// Call this only after transportMonitor goroutine has ended.
 	defer func() {
-		minConnectTimeout = mctBkp
+		atomic.StoreInt64((*int64)(&minConnectTimeout), int64(mctBkp))
 	}()
 	defer leakcheck.Check(t)
-	minConnectTimeout = time.Millisecond * 500
+	atomic.StoreInt64((*int64)(&minConnectTimeout), int64(time.Millisecond*500))
 	server, err := net.Listen("tcp", "localhost:0")
 	if err != nil {
 		t.Fatalf("Error while listening. Err: %v", err)

